File: adapters/prime/refine.py
"""Self-modification, observed on both paths and announced once (ADR-0158 D1~D4).

## Two paths, one fact

prime's continual-harness refinement reaches `harness_state.json` two ways, and
they have different visibility (measured, `research/2026-08-07-prime-refine-upstream-draft.md`):

1. **The host command path.** `refine` produces a `refine_complete` record on the
   same stdout as every other event. It is real, it is undocumented, and it is
   absent from the shipped RPC types — so an adapter has to know the name by
   hand. Re-measuring that the name still exists is part of the version-pin
   discipline when this package moves off v0.7.0.
2. **The kernel path.** `rlm.harness` inside the IPython kernel writes the *same
   file* with **zero** protocol output (6 runs x 37 events, `refine_complete` = 0).
   Nothing observes it except the file.

Both change what a teammate will do next, so ADR-0158 D1 says both belong in the
channel. This module makes them one announcement type with an honest `trigger`:
the second path says `observed-drift`, which claims only what we actually know —
we saw the file change — instead of claiming the agent decided something.

## Idempotency (D4)

ADR-0158 D4 says the key is `RefinementResult.id`, and the *property* it wants is
what this module delivers: the harness's own stable name for a refinement decides
the message, so a retry lands on the message that exists.

The id itself cannot be the key, and that is measured, not assumed.
`SendMessageRequest.client_msg_id` is a `Uuid` on the server, so posting
`refine_20260807041452415` answers **422** *"clientMsgId: UUID parsing failed:
invalid character: found `r` at 0"*. The key is therefore
`uuid5("refinement", <RefinementResult.id>)` — a pure function of exactly the
value D4 named, which keeps its guarantee while being a value the route can
decode. For an observed drift there is no such id, so the key is a UUIDv5 over
the **content hash of the state file**: observing the same state twice names the
same message, and only a real further change names a new one.

The in-process `emitted` set is the first line and the server's `clientMsgId`
unique index is the second. Neither alone is enough: the set forgets across a
restart, and the index cannot stop us from minting a fresh key.

## What the channel is told, and what it is not

Props carry `trigger`, `entryIds`, `refinementIds`, `scope` — ids and shape, no
content. Harness entries can quote the conversation that produced them, so
pouring them into a channel would leak a summary of everything the agent read.
The full text stays in the worker host's file; the channel gets "what kind, how
many".

`scope` is always `workspace`. The harness's own word for a cross-session write
is "global", but this adapter runs one workspace per HOME (see
`container/entrypoint.sh`), so the harness's global *is* our workspace. Passing
its word through would be a lie in the one direction that matters.
"""

from __future__ import annotations

import hashlib
import json
import os
from typing import Any

from .oort_client import OortClient, stable_key

TRIGGER_COMMAND = "command"
TRIGGER_TURN_INTERVAL = "turn_interval"
TRIGGER_COMPACT = "compact"
TRIGGER_OBSERVED_DRIFT = "observed-drift"

#: The props key ADR-0158 D2 puts the evidence under. Namespaced under `momo.`
#: like `momo.stream`, because the wire namespace is frozen (ADR-0152 D1).
REFINE_PROPS_KEY = "momo.harnessRefine"

#: Scope as the channel is allowed to read it. See the module docstring.
WORKSPACE_SCOPE = "workspace"


class HarnessObserver:
    """The disk-side half of refine auditing: hash the state file, diff the ids.

    This exists because of the kernel path. If upstream ever emits an event on
    every `HarnessState.save()` regardless of who called it, this class is the
    code that gets deleted — and the `observed-drift` trigger with it.
    """

    def __init__(self, path: str | None = None):
        self.path = path or default_harness_state_path()
        self.baseline = self.snapshot()

    def snapshot(self) -> dict[str, Any]:
        snap: dict[str, Any] = {"path": self.path, "exists": os.path.exists(self.path)}
        if not snap["exists"]:
            snap["sha256"] = None
            snap["entryIds"] = []
            snap["refinementIds"] = []
            return snap
        try:
            with open(self.path, "rb") as handle:
                raw = handle.read()
        except OSError as exc:  # the file can vanish between the check and the read
            snap.update({"exists": False, "sha256": None, "entryIds": [], "refinementIds": [], "error": str(exc)})
            return snap
        snap["sha256"] = hashlib.sha256(raw).hexdigest()
        snap["bytes"] = len(raw)
        snap["entryIds"] = []
        snap["refinementIds"] = []
        try:
            state = json.loads(raw)
        except json.JSONDecodeError:
            # A half-written file is a real observation, not a crash: record that
            # it changed and let the ids stay empty rather than dropping the
            # drift entirely.
            snap["parse"] = "failed"
            return snap
        entries = state.get("entries") or {}
        if isinstance(entries, dict):
            snap["entryIds"] = sorted(
                str(entry_id)
                for kind in entries.values()
                if isinstance(kind, dict)
                for entry_id in kind
            )
        refinements = state.get("refinements") or []
        if isinstance(refinements, list):
            snap["refinementIds"] = [str(item.get("id")) for item in refinements if isinstance(item, dict)]
        return snap

    def drift(self) -> dict[str, Any] | None:
        """The change since the last accepted baseline, or `None`.

        Reading advances nothing: [`accept`] is separate so a failed announcement
        is retried on the next turn instead of being lost because the observer
        already moved on.
        """
        current = self.snapshot()
        if current.get("sha256") == self.baseline.get("sha256"):
            return None
        before_entries = set(self.baseline.get("entryIds") or [])
        before_refinements = set(self.baseline.get("refinementIds") or [])
        return {
            "before": self.baseline,
            "after": current,
            "newEntryIds": sorted(set(current.get("entryIds") or []) - before_entries),
            "newRefinementIds": [
                item for item in (current.get("refinementIds") or []) if item not in before_refinements
            ],
        }

    def accept(self) -> dict[str, Any]:
        """Move the baseline to what is on disk now, and return it.

        Called after a `refine_complete` announcement too — otherwise the file
        change that refinement just made would be re-announced a second time as
        an `observed-drift`, which is the same fact wearing a worse name.
        """
        self.baseline = self.snapshot()
        return self.baseline


def default_harness_state_path() -> str:
    """Where prime keeps global harness state, resolved the way prime resolves it.

    `$PRIME_AGENT_CODING_AGENT_DIR` else `$HOME/.prime/agent` (`dist/config.js`
    `getAgentDir`, and `rlm/harness.py:_agent_dir` on the kernel side). Both
    halves matter: the kernel writes through the Python spelling, and if this
    adapter watched a different path it would observe nothing at all.
    """
    agent_dir = os.environ.get("PRIME_AGENT_CODING_AGENT_DIR") or os.path.join(
        os.path.expanduser("~"), ".prime", "agent"
    )
    return os.path.join(agent_dir, "harness", "harness_state.json")


def default_client_msg_id(refinement_id: str | None, digest: str | None) -> str:
    """The announcement's idempotency key (D4), as a value the route can decode.

    Both branches are pure functions of something the harness owns: the
    refinement's own id, or the content hash of the state file it wrote. Nothing
    here consults a clock or a counter, which is the whole property — two
    processes observing the same fact name the same message.
    """
    if refinement_id:
        return stable_key("refinement", refinement_id)
    return stable_key("drift", digest or "")


def refine_body(handle: str, entry_count: int, trigger: str) -> str:
    """The human sentence. Evidence lives in props; this is what a person reads.

    Same discipline as the approval card: the body is a plain statement, and the
    ids that back it are one layer down for whoever wants them.
    """
    who = handle or "에이전트"
    if trigger == TRIGGER_OBSERVED_DRIFT:
        if entry_count > 0:
            return f"{who}의 작업 방식이 갱신된 것을 확인했습니다 (항목 {entry_count}건)"
        return f"{who}의 작업 방식이 갱신된 것을 확인했습니다"
    if entry_count > 0:
        return f"{who}이 자기 작업 방식을 갱신했습니다 (항목 {entry_count}건)"
    return f"{who}이 자기 작업 방식을 갱신했습니다"


class RefineAnnouncer:
    """Turns an observed refinement into exactly one `system` message.

    Type `system` rather than a new message type is ADR-0158 D2: the timeline
    already has a quiet line for "something happened that is not someone
    talking", and adding a type costs every client a branch. The props key is
    what a future filter would split on if the demand ever shows up.
    """

    def __init__(
        self,
        client: OortClient,
        *,
        agent_handle: str = "",
        observer: HarnessObserver | None = None,
        key_factory=None,
    ):
        self.client = client
        self.agent_handle = agent_handle
        self.observer = observer
        # Seam for the red proof: swapping in a per-call unique key reproduces
        # the duplicate announcement that D4 exists to prevent. Production code
        # never passes this.
        self._key_factory = key_factory or default_client_msg_id
        self.emitted: set[str] = set()
        self.announcements: list[dict[str, Any]] = []

    def announce_refine_complete(
        self, result: dict[str, Any], *, trigger: str = TRIGGER_COMMAND
    ) -> dict[str, Any] | None:
        """The RPC path — `refine_complete` carried a `RefinementResult`."""
        refinement_id = str(result.get("id") or "") or None
        applied = [edit for edit in (result.get("appliedEdits") or []) if isinstance(edit, dict)]
        entry_ids = sorted({str(edit.get("id")) for edit in applied if edit.get("id")})
        refinement_ids = [refinement_id] if refinement_id else []
        digest = None
        if self.observer is not None:
            digest = self.observer.snapshot().get("sha256")
        announcement = self._announce(
            client_msg_id=self._key_factory(refinement_id, digest),
            trigger=trigger,
            entry_ids=entry_ids,
            refinement_ids=refinement_ids,
        )
        if self.observer is not None:
            # The refinement just wrote the file; adopt that as the new baseline
            # so it is not announced twice under two triggers.
            self.observer.accept()
        return announcement

    def announce_observed_drift(self, drift: dict[str, Any]) -> dict[str, Any] | None:
        """The kernel path — nothing said anything, the file simply changed."""
        after = drift.get("after") or {}
        digest = after.get("sha256")
        announcement = self._announce(
            client_msg_id=self._key_factory(None, digest),
            trigger=TRIGGER_OBSERVED_DRIFT,
            entry_ids=list(drift.get("newEntryIds") or []),
            refinement_ids=list(drift.get("newRefinementIds") or []),
        )
        if self.observer is not None:
            self.observer.accept()
        return announcement

    def _announce(
        self,
        *,
        client_msg_id: str,
        trigger: str,
        entry_ids: list[str],
        refinement_ids: list[str],
    ) -> dict[str, Any] | None:
        if client_msg_id in self.emitted:
            # Second line of defence is the server's unique index; this one keeps
            # a restart-free process from even asking.
            return None
        evidence = {
            "trigger": trigger,
            "entryIds": entry_ids,
            "refinementIds": refinement_ids,
            "scope": WORKSPACE_SCOPE,
        }
        result = self.client.post_message(
            client_msg_id=client_msg_id,
            message_type="system",
            body=refine_body(self.agent_handle, len(entry_ids), trigger),
            props={"harness": "prime-agent", REFINE_PROPS_KEY: evidence},
        )
        self.emitted.add(client_msg_id)
        record = {"clientMsgId": client_msg_id, "evidence": evidence, "result": result}
        self.announcements.append(record)
        return record

File: server-rust/crates/momo-messaging/src/refine.rs
//! ADR-0158 D1~D4 — **the agent said it changed itself**, as one channel line.
//!
//! A prime harness rewrites its own instructions on a timer
//! (`getAutoRefineSettings()` is `enabled ?? true`, `turnInterval` 25,
//! `cooldownMs` 20 minutes — 실측, `research/2026-08-07-prime-refine-upstream-draft.md`
//! §1.1). Nothing about that reaches a channel today, and the whole argument of
//! D1 is that in a product where an agent is a first-class member, "a colleague
//! changed how they work" is a fact the team is owed rather than a line in one
//! host's log.
//!
//! ## What this module is, and what it deliberately is not
//!
//! It is a **shape**: a validated block, a server-owned props key, and one
//! derived idempotency key. It is not a new message type (D2 — `system` is
//! reused, so no client learns a frame), not a second write path (the
//! announcement goes through `send_message_with_mentions_in_tx` like every other
//! send), and not a rollback affordance (D3 — `rollbackId` is recorded and
//! nothing is promised about it).
//!
//! ## Why it is not the `momo.stream` contract wearing a different hat
//!
//! The sketch's §2.1 table is worth keeping in view, because the two look alike
//! from a distance and behave oppositely:
//!
//! | | a stream slice (#1152) | a refinement |
//! |---|---|---|
//! | what changed | one message's body | the agent's **later behaviour** — no message changes |
//! | frequency | 17 per turn (실측) | gated by 25 turns *and* a 20-minute cooldown |
//! | `seq` | must **not** be consumed — 17 slices must not read as 17 unreads | **must** be consumed — an audit you cannot scroll back to is not an audit |
//! | rewind guard | `momo.stream.rev` monotonicity | none needed; each refinement is its own event |
//!
//! Routing a refinement through the stream contract would feed refinement
//! revisions into the staleness guard and freeze every later slice of that
//! message as stale. The two contracts are orthogonal, and this module exists so
//! that they stay that way.

use serde_json::{json, Map, Value};
use uuid::Uuid;

/// The server-owned props key a refinement announcement writes into (D2).
///
/// Server-owned for the same reason [`crate::STREAM_PROPS_KEY`] is: the route
/// strips it from client-supplied props so this crate is its only writer. The
/// difference from the stream marker is *why* — nothing reads this key back to
/// make a decision, so the risk is not a poisoned guard but a **forged claim**.
/// A props map is otherwise the producer's own dictionary, and a client that
/// could spell this key by hand could put "김인턴이 자기 작업 방식을 갱신했습니다"
/// under a namespace that says the server vouched for it.
pub const HARNESS_REFINE_PROPS_KEY: &str = "momo.harnessRefine";

/// The `scope` a refinement announcement may claim, and the only one (§2.2).
///
/// The harness's own `scope` for a global refinement is `"global"`, and
/// repeating that word here would be a lie in our vocabulary: an adapter runs
/// one `HOME` per workspace (#1162 tenancy 결론), so what the harness calls
/// global is, on our side of the boundary, exactly one workspace. A value the
/// server cannot vouch for is refused rather than translated silently, because a
/// silent translation is indistinguishable from a correct claim.
pub const HARNESS_REFINE_SCOPE: &str = "workspace";

/// The namespace the refinement idempotency key hangs off (D4).
///
/// Sixteen ASCII bytes, exactly like `tool_exec::TOOL_RESULT_NAMESPACE`, and for
/// the same auditable reason: byte 6 is `a` (`0x61`), so its version nibble is
/// `6`, while every `agent_run.id` is `uuidv7()`. No run id can collide with
/// this namespace, so the refinement key space and the tool-card key space
/// (`Uuid::new_v5(&run_id, …)`) cannot meet at their roots.
const HARNESS_REFINE_NAMESPACE: Uuid = Uuid::from_bytes(*b"momo.harnessRefi");

/// The `clientMsgId` a refinement announcement must carry — **derived, never
/// invented** (D4).
///
/// `RefinementResult.id` (`refine_20260807041452415`) is already a stable key
/// the harness assigned, which is what makes the spike's §8 complaint — "no
/// logic re-uses the same key on a retry" — disappear here for free. It is also
/// an arbitrary string, and `message.client_msg_id` is a `uuid` column
/// (`schema_v0.sql:180`, 동결층). So the string is *hashed into* a uuid rather
/// than parsed as one, which is the same move `tool_exec::result_message_id`
/// makes for a provider's `call_id` and for the same stated reason.
///
/// The determinism **is** the guarantee: two POSTs of one refinement — a retry
/// after a timeout, the RPC path and the file watcher both noticing the same
/// event — derive one key, meet the spine's `(channel, author, client_msg_id)`
/// unique index, and leave one line in the channel.
pub fn harness_refine_client_msg_id(refinement_id: &str) -> Uuid {
    Uuid::new_v5(&HARNESS_REFINE_NAMESPACE, refinement_id.as_bytes())
}

/// What set a refinement off (§2.2).
///
/// Four values and no `Other`: an announcement whose trigger this server does
/// not recognise is refused, because the sentence a reader is shown ("자기
///작업 방식을 갱신했습니다") is a claim about *why*, and a server that passed
/// an unknown word through would be vouching for a story it cannot read.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum HarnessRefineTrigger {
    /// The host asked for it (`refine` RPC command).
    Command,
    /// `turnInterval` elapsed — the on-by-default timer.
    TurnInterval,
    /// A compaction pass carried it.
    Compact,
    /// **The kernel path.** `rlm.harness` wrote `harness_state.json` directly and
    /// the protocol said nothing at all (실측: 6 runs × 37 events,
    /// `refine_complete` = 0). The adapter noticed by hashing the file, so this
    /// value states exactly that and no more — "we saw the file change", never
    /// "the agent decided X". Deleting this variant when upstream emits an event
    /// for kernel writes is the intended end of it.
    ObservedDrift,
}

impl HarnessRefineTrigger {
    /// The wire spelling. The casing is inconsistent (`turn_interval` snake,
    /// `observed-drift` kebab) because the sketch's table is, and the sketch is
    /// what the adapter was written against — tidying it here would be a rename
    /// nobody asked for that lands as a 400 on the other side.
    pub fn wire(self) -> &'static str {
        match self {
            HarnessRefineTrigger::Command => "command",
            HarnessRefineTrigger::TurnInterval => "turn_interval",
            HarnessRefineTrigger::Compact => "compact",
            HarnessRefineTrigger::ObservedDrift => "observed-drift",
        }
    }

    pub fn from_wire(value: &str) -> Option<HarnessRefineTrigger> {
        match value {
            "command" => Some(HarnessRefineTrigger::Command),
            "turn_interval" => Some(HarnessRefineTrigger::TurnInterval),
            "compact" => Some(HarnessRefineTrigger::Compact),
            "observed-drift" => Some(HarnessRefineTrigger::ObservedDrift),
            _ => None,
        }
    }

    /// Every spelling, for the refusal sentence and the openapi enum.
    pub const ALL: [HarnessRefineTrigger; 4] = [
        HarnessRefineTrigger::Command,
        HarnessRefineTrigger::TurnInterval,
        HarnessRefineTrigger::Compact,
        HarnessRefineTrigger::ObservedDrift,
    ];
}

/// One harness entry a refinement touched — **its name and its kind, never its
/// text** (§2.2).
///
/// The rule is a disclosure rule, not a size one. Harness memories can quote the
/// conversation that produced them, so a `before`/`after` pair in a channel is a
/// summary of that conversation posted to everyone in the room. The full record
/// stays on the adapter host, where `RefinementResult.appliedEdits` already
/// keeps it; the channel gets "무엇이 몇 건".
///
/// The wire type is `deny_unknown_fields`, which is what makes this rule
/// mechanical rather than aspirational: a producer that adds `before` gets a
/// 400, not a silently trimmed field it will believe was delivered.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct HarnessRefineEdit {
    /// `create` / `update` / `delete` as the harness reports it. Free text
    /// rather than an enum: this mirrors an upstream vocabulary that is
    /// undocumented (the whole subject of the §1 issue draft), and a closed set
    /// here would turn a new upstream verb into a 400 on an announcement that is
    /// otherwise perfectly true.
    pub action: String,
    /// `memory` / `instruction` / … — same reasoning as `action`.
    pub kind: String,
    /// The harness's own entry id (실측: `oort-refine-probe`).
    pub id: String,
}

/// A validated refinement announcement, ready to become props.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct HarnessRefine {
    pub refinement_id: String,
    pub trigger: HarnessRefineTrigger,
    pub edits: Vec<HarnessRefineEdit>,
    pub summary: Option<String>,
    /// D3 — **recorded, not exposed.** `refine` can be rewound by `rollbackId`
    /// (`RefineOptions`), and v0 deliberately ships no channel affordance for
    /// that: what a rewind should look like to a room is an unanswered UX
    /// question (§2.4), and shipping a half-answer would be harder to withdraw
    /// than to add. Keeping the id on the row is the cheap half — an operator
    /// answering "can this be undone, and by what id" reads the ledger.
    pub rollback_id: Option<String>,
}

/// The longest `refinementId` / edit field this server will store.
///
/// These are identifiers, not prose: the measured ones are 23 and 17 characters
/// (`refine_20260807041452415`, `oort-refine-probe`). The cap exists so that a
/// producer cannot use an id field as the content channel §2.2 closed.
pub const HARNESS_REFINE_ID_MAX_CHARS: usize = 200;

/// The longest human `summary` a refinement may carry.
///
/// Bounded for the disclosure reason, not a storage one — an unbounded summary
/// is exactly the "요약본 유출" §2.2 refuses, re-opened under a friendlier name.
pub const HARNESS_REFINE_SUMMARY_MAX_CHARS: usize = 500;

/// The most edits one announcement may enumerate.
///
/// A refinement that touched more entries than this is still announced; the
/// count simply stops being an itemised list, which is the failure mode that
/// costs the least. (No measured run came close — the spike's applied edits
/// numbered 1.)
pub const HARNESS_REFINE_EDITS_MAX: usize = 50;

/// Why a refinement announcement was refused **at the acceptance point**.
///
/// All of them are the caller's to fix, and all of them are 400 at the route.
/// Refusing here rather than dropping the block is the #1183 discipline applied
/// one layer up: a producer that believes it announced a self-modification, and
/// did not, has no way to discover the difference.
#[derive(Debug, Clone, PartialEq, Eq, thiserror::Error)]
pub enum HarnessRefineInvalid {
    #[error("harnessRefine requires type \"system\" (ADR-0158 D2)")]
    NotSystemMessage,
    #[error("harnessRefine requires a human-readable body — the reason belongs in props, the sentence in the channel")]
    MissingBody,
    #[error("harnessRefine.refinementId must be a non-empty id of at most {HARNESS_REFINE_ID_MAX_CHARS} characters")]
    RefinementId,
    #[error(
        "harnessRefine.trigger must be one of: command, turn_interval, compact, observed-drift"
    )]
    Trigger,
    #[error("harnessRefine.scope must be \"{HARNESS_REFINE_SCOPE}\" — an adapter runs one HOME per workspace, so the harness's \"global\" is this workspace and nothing wider")]
    Scope,
    #[error("harnessRefine.edits[].action/kind/id must each be a non-empty id of at most {HARNESS_REFINE_ID_MAX_CHARS} characters")]
    Edit,
    #[error("harnessRefine.edits may name at most {HARNESS_REFINE_EDITS_MAX} entries")]
    TooManyEdits,
    #[error("harnessRefine.summary must be at most {HARNESS_REFINE_SUMMARY_MAX_CHARS} characters — the harness text stays on the adapter host")]
    Summary,
    /// The idempotency key is derived, so a caller that supplies a different one
    /// is told which one this refinement has. Naming the value is what keeps the
    /// refusal actionable rather than a riddle.
    #[error("clientMsgId for this refinement must be {expected} — the key is derived from refinementId so a retry cannot announce twice (ADR-0158 D4)")]
    ClientMsgId { expected: Uuid },
}

impl HarnessRefine {
    /// The `props["momo.harnessRefine"]` object this announcement stores.
    ///
    /// `summary`/`rollbackId` are **absent** rather than `null` when unset, the
    /// same shape rule ADR-0155's `outcome` follows: a producer written before a
    /// field existed sends what it always sent, and a reader cannot tell the two
    /// apart from a field that was explicitly cleared.
    pub fn props(&self) -> Value {
        let mut object = Map::new();
        object.insert(
            "refinementId".to_string(),
            Value::String(self.refinement_id.clone()),
        );
        object.insert(
            "trigger".to_string(),
            Value::String(self.trigger.wire().to_string()),
        );
        // Written by the server, not copied from the request — the request's
        // value was only ever a claim this server had to agree with.
        object.insert(
            "scope".to_string(),
            Value::String(HARNESS_REFINE_SCOPE.to_string()),
        );
        object.insert(
            "edits".to_string(),
            Value::Array(
                self.edits
                    .iter()
                    .map(|edit| json!({ "action": edit.action, "kind": edit.kind, "id": edit.id }))
                    .collect(),
            ),
        );
        if let Some(summary) = self.summary.as_ref() {
            object.insert("summary".to_string(), Value::String(summary.clone()));
        }
        if let Some(rollback_id) = self.rollback_id.as_ref() {
            object.insert("rollbackId".to_string(), Value::String(rollback_id.clone()));
        }
        Value::Object(object)
    }

    /// The derived key this announcement must be sent under (D4).
    pub fn client_msg_id(&self) -> Uuid {
        harness_refine_client_msg_id(&self.refinement_id)
    }
}

/// One id-shaped field's check. Counted in characters, not bytes, because a cap
/// a Korean id trips at a third of its stated length is not the cap that was
/// documented.
fn is_valid_id(value: &str) -> bool {
    !value.trim().is_empty() && value.chars().count() <= HARNESS_REFINE_ID_MAX_CHARS
}

/// Validate the wire block's *values* — the half `deny_unknown_fields` cannot do.
///
/// Serde refuses fields nobody declared; this refuses declared fields whose
/// values the server would otherwise be repeating on the producer's word. Split
/// this way because the two failures are different sentences: "you sent a key I
/// do not know" and "the scope you claimed is not one I can vouch for".
pub fn validate_harness_refine(
    refinement_id: &str,
    trigger: &str,
    scope: &str,
    edits: &[(String, String, String)],
    summary: Option<&str>,
    rollback_id: Option<&str>,
) -> Result<HarnessRefine, HarnessRefineInvalid> {
    if !is_valid_id(refinement_id) {
        return Err(HarnessRefineInvalid::RefinementId);
    }
    let trigger = HarnessRefineTrigger::from_wire(trigger).ok_or(HarnessRefineInvalid::Trigger)?;
    if scope != HARNESS_REFINE_SCOPE {
        return Err(HarnessRefineInvalid::Scope);
    }
    if edits.len() > HARNESS_REFINE_EDITS_MAX {
        return Err(HarnessRefineInvalid::TooManyEdits);
    }
    let mut checked = Vec::with_capacity(edits.len());
    for (action, kind, id) in edits {
        if !is_valid_id(action) || !is_valid_id(kind) || !is_valid_id(id) {
            return Err(HarnessRefineInvalid::Edit);
        }
        checked.push(HarnessRefineEdit {
            action: action.clone(),
            kind: kind.clone(),
            id: id.clone(),
        });
    }
    let summary = match summary {
        None => None,
        Some(summary) => {
            if summary.chars().count() > HARNESS_REFINE_SUMMARY_MAX_CHARS {
                return Err(HarnessRefineInvalid::Summary);
            }
            Some(summary.to_string())
        }
    };
    let rollback_id = match rollback_id {
        None => None,
        Some(rollback_id) => {
            if !is_valid_id(rollback_id) {
                return Err(HarnessRefineInvalid::RefinementId);
            }
            Some(rollback_id.to_string())
        }
    };
    Ok(HarnessRefine {
        refinement_id: refinement_id.to_string(),
        trigger,
        edits: checked,
        summary,
        rollback_id,
    })
}

/// The send's props with the refinement block stamped in.
///
/// Merged into whatever the producer already sent (`{"harness": "prime-agent"}`
/// in the sketch), never replacing it — the same merge
/// `message::opening_stream_input` makes, so an announcement can also carry the
/// ordinary producer props every other message of that adapter carries.
pub fn harness_refine_input_props(props: Value, refine: &HarnessRefine) -> Value {
    let mut object = match props {
        Value::Object(object) => object,
        _ => Map::new(),
    };
    object.insert(HARNESS_REFINE_PROPS_KEY.to_string(), refine.props());
    Value::Object(object)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn valid() -> Result<HarnessRefine, HarnessRefineInvalid> {
        validate_harness_refine(
            "refine_20260807041452415",
            "command",
            "workspace",
            &[(
                "create".to_string(),
                "memory".to_string(),
                "oort-refine-probe".to_string(),
            )],
            Some("기억 1건 추가"),
            None,
        )
    }

    /// The measured shape from `run_spike.sh refine` parses, field for field.
    /// Written against the 실측 values rather than convenient ones so that a
    /// rename upstream shows up here rather than in production.
    #[test]
    fn the_measured_refinement_result_is_accepted_verbatim() {
        let refine = valid().expect("the measured shape is valid");
        assert_eq!(refine.refinement_id, "refine_20260807041452415");
        assert_eq!(refine.trigger, HarnessRefineTrigger::Command);
        assert_eq!(refine.edits.len(), 1);
        assert_eq!(refine.edits[0].id, "oort-refine-probe");

        let props = refine.props();
        assert_eq!(props["scope"], json!("workspace"));
        assert_eq!(props["trigger"], json!("command"));
        assert_eq!(props["edits"][0]["kind"], json!("memory"));
        assert!(
            props.get("rollbackId").is_none(),
            "absent, not null — a producer that never had the field means what it always meant"
        );
    }

    /// **D4, and the reason this module owns a namespace.** The key is a pure
    /// function of the refinement id: an RPC-sourced announcement and a
    /// file-watcher-sourced one for the same refinement collapse in the spine's
    /// unique index instead of double-announcing.
    #[test]
    fn the_idempotency_key_is_a_pure_function_of_the_refinement_id() {
        let once = harness_refine_client_msg_id("refine_20260807041452415");
        let again = harness_refine_client_msg_id("refine_20260807041452415");
        assert_eq!(once, again);
        assert_ne!(
            once,
            harness_refine_client_msg_id("refine_20260807041452416")
        );
        assert_eq!(valid().expect("valid").client_msg_id(), once);
    }

    /// The namespace is pinned by value, because an adapter in another language
    /// re-derives this key and a "tidy-up" of the constant would silently split
    /// the key space in two — every retry after it becoming a second
    /// announcement of a refinement that happened once.
    #[test]
    fn the_namespace_is_pinned_and_cannot_collide_with_a_run_id() {
        assert_eq!(
            HARNESS_REFINE_NAMESPACE.to_string(),
            "6d6f6d6f-2e68-6172-6e65-737352656669",
            "16 ASCII bytes of `momo.harnessRefi` — the value an adapter in \
             another language pastes"
        );
        // Byte 6's high nibble is the version field. `agent_run.id` is uuidv7;
        // this is 6, so no run id can ever equal the namespace.
        assert_eq!(HARNESS_REFINE_NAMESPACE.as_bytes()[6] >> 4, 6);
    }

    /// §2.2's disclosure rule, asserted as a refusal rather than a comment.
    #[test]
    fn a_scope_the_server_cannot_vouch_for_is_refused() {
        let global = validate_harness_refine("refine_1", "command", "global", &[], None, None);
        assert_eq!(global, Err(HarnessRefineInvalid::Scope));
    }

    /// An unknown trigger is a story this server cannot read, so it does not
    /// repeat it.
    #[test]
    fn an_unknown_trigger_is_refused_rather_than_passed_through() {
        assert_eq!(
            validate_harness_refine("refine_1", "vibes", "workspace", &[], None, None),
            Err(HarnessRefineInvalid::Trigger)
        );
        for trigger in HarnessRefineTrigger::ALL {
            assert_eq!(
                HarnessRefineTrigger::from_wire(trigger.wire()),
                Some(trigger),
                "every spelling this server advertises must round-trip"
            );
        }
    }

    /// The caps are disclosure guards, so they are asserted at the boundary that
    /// matters — one character over is a refusal, not a truncation.
    #[test]
    fn an_oversized_summary_is_refused_not_truncated() {
        let long = "가".repeat(HARNESS_REFINE_SUMMARY_MAX_CHARS + 1);
        assert_eq!(
            validate_harness_refine("refine_1", "command", "workspace", &[], Some(&long), None),
            Err(HarnessRefineInvalid::Summary)
        );
        let at_limit = "가".repeat(HARNESS_REFINE_SUMMARY_MAX_CHARS);
        assert!(
            validate_harness_refine(
                "refine_1",
                "command",
                "workspace",
                &[],
                Some(&at_limit),
                None
            )
            .is_ok(),
            "counted in characters, not bytes"
        );
    }

    /// D3 — stored when sent, and it does not change the shape when it is not.
    #[test]
    fn a_rollback_id_is_recorded_without_being_promised() {
        let refine = validate_harness_refine(
            "refine_1",
            "command",
            "workspace",
            &[],
            None,
            Some("rollback_20260807"),
        )
        .expect("valid");
        assert_eq!(refine.props()["rollbackId"], json!("rollback_20260807"));
    }

    /// The block joins the producer's own props instead of replacing them.
    #[test]
    fn the_refine_block_joins_the_props_a_send_already_carried() {
        let refine = valid().expect("valid");
        let props = harness_refine_input_props(json!({ "harness": "prime-agent" }), &refine);
        assert_eq!(props["harness"], json!("prime-agent"));
        assert_eq!(
            props[HARNESS_REFINE_PROPS_KEY]["refinementId"],
            json!("refine_20260807041452415")
        );
    }
}
